### ISSUE TITLE
Add padding feature in MaxPooling and MeanPooling layers.

### ISSUE THREAD

**Issue by walragatver** [mlpack/mlpack#2119]
Currently, our pooling layers doesn't have support for padding. However, that would be a nice feature to add. 

The following are the basic steps to do this
1. Go through the following links and implementation of `mean_pooling` and `max_pooling` layers in `/src/mlpack/methods/ann/layer` folder to understand what is the requirement. 

https://arxiv.org/pdf/1603.07285.pdf
https://keras.io/layers/pooling/

2. We have padding support for the `Convolution` layers and in order to add this feature, we can refer to the implementation of `Convolution`, `AtrousConvolution` layers in `/src/mlpack/methods/ann/layer/convolution.hpp` and `/src/mlpack/methods/ann/layer/atrous_convolution.hpp`.

3. Check out how the Padding layer is used for the padding functionality in `Convolution` layers. 

4. Implementation.

`Note`: Tensorflow uses a different approach to the `same` padding in pooling layers. We follow the definition that `Same` padding is padding done such that output size is the same as input size. 
See https://stackoverflow.com/questions/37674306/what-is-the-difference-between-same-and-valid-padding-in-tf-nn-max-pool-of-t for more info on it.

**Comment by walragatver** [comment]
Hi @zoq, @sreenikSS, let me know what you think of the note I have added at the end.

**Comment by janish1999** [comment]
Hey @walragatver ,
     I'm a beginner with some knowledge of deep learning. Shall I take up this issue?

**Comment by walragatver** [comment]
@janish1999, Feel free to go ahead.

**Comment by sreenikSS** [comment]
@walragatver you are right with that info about Tensorflow. Although the output shape of the SAME padding is is equal to the input shape, the actual implementation seems to be slightly different.
Strangely enough, I found some info in a [pytorch thread](https://github.com/pytorch/pytorch/issues/3867).

**Comment by walragatver** [comment]
@sreenikSS, I hope you would not mind if we follow that 'SAME' padding means output size is same as input. Our current existing code support the given logic. If we wish to make our 'same' padding equivalent to 'tensorflow' one then we will need to change our code. Let me know what you think. 
Also I digged quite a lot and found that various libraries have thier own way of implementation. Even keras have mentioned about it. `same` padding is quite inconsistent across their backends. So considering this I think it would be better if stick to normal definition of same padding let me know what you think.
https://github.com/keras-team/keras/pull/9473#issuecomment-372166860

**Comment by sreenikSS** [comment]
>  I hope you would not mind if we follow that 'SAME' padding means output size is same as input. Our current existing code support the given logic. If we wish to make our 'same' padding equivalent to 'tensorflow' one then we will need to change our code.

@walragatver I have no issue with the implementation of the `same` padding we currently have. What you have mentioned regarding the inconsistency across several backends makes it even more cumbersome to find a standard definition other than the fact that the input and output dimensions remain the same (or at least tend to).

To summarize, I think what we have is sufficient.

**Comment by AbhinavTalari** [comment]
Hello , I'm a beginner, Can i Take up this issue and try?

**Comment by rcurtin** [comment]
> > I hope you would not mind if we follow that 'SAME' padding means output size is same as input. Our current existing code support the given logic. If we wish to make our 'same' padding equivalent to 'tensorflow' one then we will need to change our code.
> 
> @walragatver I have no issue with the implementation of the `same` padding we currently have. What you have mentioned regarding the inconsistency across several backends makes it even more cumbersome to find a standard definition other than the fact that the input and output dimensions remain the same (or at least tend to).
> 
> To summarize, I think what we have is sufficient.

Agreed, from my perspective I think that's just fine.  The key is just making sure that the functionality is well documented so that users can understand quickly and easily that it's not exactly the same as TF. :)

**Comment by walragatver** [comment]
@janish1999, @AbhinavTalari and everyone else
There's no need to claim the issue---just work on it, and once you have a working solution, open a PR and someone will review it when they have a chance. :)

**Comment by sainikhit2k** [comment]
Is anyone working on it ? If not i can work on this

**Comment by zoq** [comment]
Hello @sainikhit2k sorry for the slow response, looks like everything is already taken up, at least there is an open PR for each. If you are interested you can review one of them, but don't feel obligated.

**Comment by Vishal19111999** [comment]
@zoq is this issue solved? Can I start working on it?

**Comment by kartikdutt18** [comment]
Hi @Vishal19111999, sorry for the slow response, I think there is a PR open by @tejasvi. Have a look at #2127 .

**Comment by kartikdutt18** [comment]
Keep Open.

**Comment by ShikharJ** [comment]
> Keep Open.

The hero we deserve, but not the one we need in this situation :laughing:

**Comment by kartikdutt18** [comment]
Hey @ShikharJ, if you get a chance could you please take a look at the PR open by @tejasvi (#2127). I think it is ready (aside from adding it to History.md).
